# Support Vector Machine(Classification)

Definition:

- 支持向量机分类器：根据训练样本的分布搜索所有可能的线性分类器中最佳的一个（因此它不一定是“最佳的”，因为最佳还要看测试集）。


- 支持向量：决定直线位置（线性分类器）的是不同空间间隔最小的不同分类的数据点。这种真正帮助决策最优线性分类模型的数据点才是支持向量。注意点也是向量。**相比Logistic，它并未考虑全部数据点**。

# 手写体数码图像识别

The first step: read the data.

In [2]:
# from Sklearn import 手写体数字加载器
from sklearn.datasets import load_digits

# save the data
digits = load_digits()

# see the shape
print digits.data.shape
# cannot ignore the 'data.', which is an attribute error, though

(1797L, 64L)


* 这个意思是图像数据有1797条，每幅图由8\*8=64的像素矩阵表示。


* 一般处理这种像素矩阵时，逐行首尾拼接成1维像素特征向量。因此会损失结构信息。不过下面的SVM学不了结构信息，所以就不管了。

The Second step: split the data into two parts.

In [3]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size = 0.25, random_state = 33)

print y_train.shape
print y_test.shape

(1347L,)
(450L,)


C:\Users\Administrator\Anaconda2\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


The third step: use SVM.

In [4]:
# import Standardization mode
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
X_train = ss.fit_transform(X_train)
X_test = ss.fit_transform(X_test)

# import LinearSVC
from sklearn.svm import LinearSVC
lsvc = LinearSVC()

# train the model
lsvc.fit(X_train, y_train)

# predict
y_predict = lsvc.predict(X_test)

The fourth step: Accuracy, Recall, Precison and F1 score.

In [5]:
# default evaluation
print 'The Accuracy of Linear SVC is', lsvc.score(X_test, y_test)

# use classification_report in sklearn.metrics
from sklearn.metrics import classification_report
print classification_report(y_test, y_predict, target_names = digits.target_names.astype(str))

The Accuracy of Linear SVC is 0.9488888888888889
             precision    recall  f1-score   support

          0       0.92      0.97      0.94        35
          1       0.95      0.98      0.96        54
          2       0.98      1.00      0.99        44
          3       0.93      0.93      0.93        46
          4       0.97      1.00      0.99        35
          5       0.94      0.94      0.94        48
          6       0.96      0.98      0.97        51
          7       0.90      1.00      0.95        35
          8       0.98      0.83      0.90        58
          9       0.95      0.91      0.93        44

avg / total       0.95      0.95      0.95       450



* 注意这里本来那些指标是用于二分类任务的，这里则对于分类目标类别较多时，逐一评估某类别的三个性能指标，创造多个二分类任务。


* SVM选择的样本是支持向量，节省学习需要的内存，提高了预测性能；但是付出了更多的CPU和计算时间。